control_apps/src/control_mngr.rs

```diff
@@ -0,0 +1,292 @@
+        debug!("Attempting transition to {:}", ControlMode::Safe);
+        let previous_mode = self.mode;
+        // unlock the valves if they are locked to force the next step!
+        self.valve_vent.unlock();
+        self.valve_dump.unlock();
+        // execute mode transition
+        self.mode = ControlMode::Safe;
+        warn!("CtrlMode transition: {:} --> {:}", previous_mode, self.mode);
+        // lock the vent valve closed
+        self.valve_vent.set_pwm(0.0);
+        self.valve_vent.lock();
+        // lock the dump valve closed
+        self.valve_dump.set_pwm(0.0);
+        self.valve_dump.lock();
+    }
+
+    fn idle(&mut self) {
+        // enter mode for holding altitude, nothing to do
+        debug!("Attempting transition to {:}", ControlMode::Idle);
+        let previous_mode = self.mode;
+        if previous_mode != ControlMode::Safe
+            && previous_mode != ControlMode::Abort
+            && previous_mode != ControlMode::Init
+        {
+            // unlock the valves
+            self.valve_vent.unlock();
+            self.valve_dump.unlock();
+            self.mode = ControlMode::Idle;
+            info!("CtrlMode transition: {:} --> {:}", previous_mode, self.mode);
+        } else {
+            warn!(
+                "CtrlMode transition ({:} --> {:}) not allowed! Ignoring...",
+                previous_mode,
+                ControlMode::Idle
+            );
+        }
+    }
+
+    fn vent(&mut self) {
+        // enter mode used for lowering altitude
+        debug!("Attempting transition to {:}", ControlMode::Vent);
+        let previous_mode = self.mode;
+        if previous_mode == ControlMode::Idle || previous_mode == ControlMode::Dump {
+            // stop dumping -- can only do one at a time!
+            self.valve_dump.set_pwm(0.0);
+            self.valve_dump.lock();
+            // enable venting
+            self.valve_vent.unlock();
+            self.mode = ControlMode::Vent;
+            info!("CtrlMode transition: {:} --> {:}", previous_mode, self.mode);
+        } else {
+            warn!(
+                "CtrlMode transition ({:} --> {:}) not allowed! Ignoring...",
+                previous_mode,
+                ControlMode::Idle
+            );
+        }
+    }
+
+    fn dump(&mut self) {
+        // enter mode used for raising altitude
+        debug!("Attempting transition to {:}", ControlMode::Dump);
+        let previous_mode = self.mode;
+        if previous_mode == ControlMode::Idle || previous_mode == ControlMode::Vent {
+            // stop vemtomg -- can only do one at a time!
+            self.valve_vent.set_pwm(0.0);
+            self.valve_vent.lock();
+            // enable dumping
+            self.valve_dump.unlock();
+            self.mode = ControlMode::Dump;
+            info!("CtrlMode transition: {:} --> {:}", previous_mode, self.mode);
+        } else {
+            warn!(
+                "CtrlMode transition ({:} --> {:}) not allowed! Ignoring...",
```

Comment:
Do you nominally expect to hit these else cases?  It seems like these could be prevented by just guarding when you can call dump based on the current state in the `update` function.

Reply:
Nominally no, this code would not be reached. But if it does, I want to know it tried you know?